File: src/main/java/ru/arc/autobuild/Construction.java
package ru.arc.autobuild;

import ru.arc.ARC;
import ru.arc.configs.Config;
import ru.arc.configs.ConfigManager;
import ru.arc.hooks.HookRegistry;
import ru.arc.hooks.citizens.CitizensHook;
import ru.arc.util.ParticleManager;
import ru.arc.util.Utils;
import com.destroystokyo.paper.ParticleBuilder;
import com.sk89q.worldedit.bukkit.BukkitAdapter;
import com.sk89q.worldedit.math.BlockVector3;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.Particle;
import org.bukkit.block.Barrel;
import org.bukkit.block.Block;
import org.bukkit.block.Chest;
import org.bukkit.block.data.BlockData;
import org.bukkit.inventory.ItemStack;
import org.bukkit.loot.LootContext;
import org.bukkit.loot.LootTables;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicInteger;

import static org.bukkit.Material.*;
import static ru.arc.util.Utils.rotateBlockData;

public class Construction {

    final ConstructionSite site;
    private List<BlockVector3> vectors = new ArrayList<>();
    private BukkitTask buildTask;
    private BukkitTask removeNpcTask;
    AtomicInteger pointer = new AtomicInteger(-1);
    boolean lookClose = false;
    int npcId = -1;

    Set<Material> skipMaterial = Set.of(CHEST, BARREL,
            TRAPPED_CHEST, PLAYER_HEAD, PLAYER_WALL_HEAD, BEDROCK);
    Set<Material> nonDrop = Set.of(SHORT_GRASS, TALL_GRASS, DIRT, GRASS_BLOCK);
    Map<String, String> skins = Map.of(
            "&6Петрович", "https://minesk.in/faca74c68a104b6987bc8c11ffebb092",
            "&6Николаич", "https://minesk.in/6666ba384aa3486b88c21fa7541fb856",
            "&6Иваныч", "https://minesk.in/3ff30e8f08ae48c2abece46bbf0c09d6",
            "&6Агадиль", "https://minesk.in/e8eae58c095949de87ff9c9b5b7c17f2");

    Config config = ConfigManager.of(ARC.plugin.getDataPath(), "auto-build.yml");

    public Construction(ConstructionSite site) {
        this.site = site;

        skins = config.map("construction.npc-skins", skins);
        nonDrop = config.materialSet("construction.not-drop-materials", nonDrop);
        skipMaterial = config.materialSet("construction.skip-materials", skipMaterial);
    }

    public void startBuilding() {
        if (npcId == -1) createNpc(site.centerBlock, -1);
        if (npcId != -1 && HookRegistry.citizensHook != null && lookClose) HookRegistry.citizensHook.lookClose(npcId);
        if (removeNpcTask != null && !removeNpcTask.isCancelled()) removeNpcTask.cancel();
        new BukkitRunnable() {
            @Override
            public void run() {
                ConstructionSite.Corners corners = site.getCorners();

                final int minX = corners.corner1().x();
                final int minY = corners.corner1().y();
                final int minZ = corners.corner1().z();
                final int maxX = corners.corner2().x();
                final int maxY = corners.corner2().y();
                final int maxZ = corners.corner2().z();

                vectors.clear();
                for (int y = minY; y <= maxY; y++) {
                    for (int x = minX; x <= maxX; x++) {
                        for (int z = minZ; z <= maxZ; z++) {
                            vectors.add(BlockVector3.at(x, y, z));
                        }
                    }
                }

                buildTask = new BukkitRunnable() {
                    @Override
                    public void run() {
                        if (buildNextBlocks(config.integer("construction.blocks-per-tick", 3))) {
                            this.cancel();
                            site.finalizeBuilding();
                        }
                    }
                }.runTaskTimer(ARC.plugin, 1L, config.integer("construction.cycle-duration-ticks", 10));
            }
        }.runTaskAsynchronously(ARC.plugin);
    }

    public void cancel(int destroyNpcDelaySeconds) {
        if (buildTask != null && !buildTask.isCancelled()) buildTask.cancel();
        if (removeNpcTask != null && !removeNpcTask.isCancelled()) removeNpcTask.cancel();
        Bukkit.getScheduler().runTaskLater(ARC.plugin, this::destroyNpc, destroyNpcDelaySeconds * 20L);
    }

    public int createNpc(Location location, int seconds) {
        if (HookRegistry.citizensHook == null) return -1;
        var entry = Utils.random(skins);
        npcId = HookRegistry.citizensHook.createNpc(entry.getKey(), location.toCenterLocation());

        HookRegistry.citizensHook.addChatBubble(npcId, List.of(
                new CitizensHook.HologramLine("&6Нажмите ПКМ, чтобы начать строительство", 100)
        ));
        HookRegistry.citizensHook.setSkin(npcId, entry.getValue());
        if (seconds > 0) {
            removeNpcTask = new BukkitRunnable() {
                @Override
                public void run() {
                    HookRegistry.citizensHook.deleteNpc(npcId);
                }
            }.runTaskLater(ARC.plugin, 20L * seconds);
            HookRegistry.citizensHook.lookClose(npcId);
            lookClose = true;
        }
        return npcId;
    }


    private boolean buildNextBlocks(int blocks) {
        int count = 0;
        while (true) {
            int index = pointer.incrementAndGet();
            if (index >= vectors.size()) return true;

            BlockVector3 v = vectors.get(index);

            int yOff = site.yOffset;
            Location location = new Location(site.world,
                    site.centerBlock.x() + v.x(),
                    site.centerBlock.y() + v.y() + yOff,
                    site.centerBlock.z() + v.z());
            int fullRotation = site.fullRotation();
            BlockData data = BukkitAdapter.adapt(site.building.getBlock(v, fullRotation));
            rotateBlockData(data, fullRotation);

            Block currentBlock = site.world.getBlockAt(location);

            if (data.getMaterial() == AIR && currentBlock.getType() == AIR) continue;
            if (data.equals(currentBlock.getBlockData()) || skipMaterial.contains(currentBlock.getType()))
                continue;
            if (HookRegistry.sfHook != null && HookRegistry.sfHook.isSlimefunBlock(currentBlock)) continue;

            if (currentBlock.getType() != data.getMaterial() &&
                    !nonDrop.contains(currentBlock.getType())) giveDrop(currentBlock);

            site.world.getBlockAt(location).setBlockData(data, false);
            if (location.getBlock().getState() instanceof Chest chest) {
                LootTables.SPAWN_BONUS_CHEST.getLootTable()
                        .fillInventory(
                                chest.getInventory(),
                                ThreadLocalRandom.current(),
                                new LootContext.Builder(location).build()
                        );
            } else if (location.getBlock().getState() instanceof Barrel barrel) {
                LootTables.SPAWN_BONUS_CHEST.getLootTable()
                        .fillInventory(
                                barrel.getInventory(),
                                ThreadLocalRandom.current(),
                                new LootContext.Builder(location).build()
                        );
            }
            boolean playerNearby = site.player.getLocation().distance(site.centerBlock) < 50;
            if (count == 0 && playerNearby) playEffects(location, data);
            count++;

            if (count >= blocks) break;
        }
        return false;
    }

    private void giveDrop(Block block) {
        block.getDrops().stream()
                .map(stack -> site.player.getInventory().addItem(stack))
                .flatMap(map -> map.values().stream())
                .forEach(stack -> site.player.getWorld().dropItem(site.player.getLocation(), stack));
    }

    private void playEffects(Location location, BlockData data) {
        if (config.bool("construction.play-sounds", true))
            location.getWorld().playSound(location, data.getSoundGroup().getPlaceSound(), 1f, 1f);
        if (config.bool("construction.show-particles", false)) {
            ParticleManager.queue(new ParticleBuilder(config.particle("construction.place-particle", Particle.FLAME))
                    .count(config.integer("construction.particle-count", 5))
                    .location(location)
                    .receivers(List.of(site.player))
                    .offset(config.real("construction.particle-offset", 0.25),
                            config.real("construction.particle-offset", 0.25),
                            config.real("construction.particle-offset", 0.25))
                    .extra(0.05));
        }
        if (npcId != -1) {
            if (ThreadLocalRandom.current().nextDouble() > 0.8) HookRegistry.citizensHook.faceNpc(npcId, location);
            if (data.getMaterial().isItem())
                HookRegistry.citizensHook.setMainHand(npcId, new ItemStack(data.getMaterial()));
            HookRegistry.citizensHook.animateNpc(npcId, CitizensHook.Animation.ARM_SWING);
        }
    }

    public void destroyNpc() {
        if (HookRegistry.citizensHook == null || npcId == -1) return;
        if (removeNpcTask != null && !removeNpcTask.isCancelled()) removeNpcTask.cancel();
        HookRegistry.citizensHook.deleteNpc(npcId);
    }


    public void finishInstantly() {
        cancel(0);
        buildNextBlocks(1000000);
    }
}

File: src/main/kotlin/ru/arc/autobuild/Construction.kt
package ru.arc.autobuild

import com.destroystokyo.paper.ParticleBuilder
import com.sk89q.worldedit.bukkit.BukkitAdapter
import com.sk89q.worldedit.math.BlockVector3
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.block.data.BlockData
import org.bukkit.inventory.ItemStack
import ru.arc.core.LifecycleTaskScope
import ru.arc.core.ScheduledTask
import ru.arc.core.delayed
import ru.arc.core.ticks
import ru.arc.hooks.HookRegistry
import ru.arc.hooks.citizens.CitizensHook
import ru.arc.util.BlockUtils.rotateBlockData
import ru.arc.util.Logging.debug
import ru.arc.util.Logging.error
import ru.arc.util.ParticleManager
import ru.arc.util.RandomUtils
import java.util.concurrent.ThreadLocalRandom
import java.util.concurrent.atomic.AtomicInteger

/**
 * Applies only the structural block state from a build book. A schematic does
 * not own container inventory data and must never synthesize loot on placement.
 */
internal object ConstructionBlockPlacement {
    fun apply(block: Block, blockData: BlockData) {
        block.blockData = blockData
    }
}

/**
 * Handles the block-by-block construction process.
 *
 * Creates an NPC worker and places blocks over time with effects.
 */
internal class Construction(
    private val site: ConstructionSite,
    private val buildTasks: LifecycleTaskScope = LifecycleTaskScope(),
    private val prepareBlocks: (ConstructionSite) -> List<BlockVector3> = ::prepareBlockList,
) {
    private var removeNpcTask: ScheduledTask? = null
    private var blocks = emptyList<BlockVector3>()

    val pointer = AtomicInteger(-1)
    var lookClose = false
        private set
    var npcId = -1
        private set

    // ==================== NPC ====================

    fun createNpc(location: Location, seconds: Int): Int {
        val citizens = HookRegistry.citizensHook ?: run {
            debug("[autobuild] createNpc skipped: Citizens hook missing for player={}", site.player.name)
            return -1
        }

        val (name, skinUrl) = RandomUtils.random(BuildConfig.npcSkins)
        npcId = citizens.createNpc(name, location.toCenterLocation())

        val promptTicks = if (seconds > 0) (seconds.toLong() * 20L).coerceAtMost(Int.MAX_VALUE.toLong()).toInt() else -1
        citizens.addChatBubble(
            npcId,
            BuildConfig.npcPrompt.map { CitizensHook.HologramLine(it, promptTicks) },
        )
        citizens.setSkin(npcId, skinUrl)

        if (seconds > 0) {
            removeNpcTask =
                delayed((20L * seconds).ticks) {
                    citizens.deleteNpc(npcId)
                }
            citizens.lookClose(npcId)
            lookClose = true
        }

        return npcId
    }

    fun destroyNpc() {
        if (npcId == -1) return
        removeNpcTask?.takeIf { !it.isCancelled }?.cancel()
        HookRegistry.citizensHook?.deleteNpc(npcId)
    }

    // ==================== Building ====================

    fun startBuilding() {
        if (npcId == -1) createNpc(site.centerBlock, -1)
        if (npcId != -1 && lookClose) HookRegistry.citizensHook?.lookClose(npcId)
        removeNpcTask?.takeIf { !it.isCancelled }?.cancel()

        val token = buildTasks.token()
        buildTasks.runAsync(token) {
            try {
                val preparedBlocks = prepareBlocks(site)
                buildTasks.runSync(token) {
                    blocks = preparedBlocks
                    debug(
                        "[autobuild] startBuilding player={} building={} blocks={} npcId={}",
                        site.player.name,
                        site.building.fileName,
                        blocks.size,
                        npcId,
                    )
                    buildTasks.runTimer(token, 1L, BuildConfig.cycleDurationTicks) {
                        if (placeNextBlocks(BuildConfig.blocksPerTick)) {
                            buildTasks.close()
                            site.complete()
                        }
                    }
                }
            } catch (e: Exception) {
                error("[autobuild] startBuilding failed for player={}", site.player.name, e)
                buildTasks.runSync(token) {
                    if (site.state == ConstructionState.Building) site.cancel()
                }
            }
        }
    }

    private companion object {
        fun prepareBlockList(site: ConstructionSite): List<BlockVector3> {
            val c = site.corners
            val prepared = mutableListOf<BlockVector3>()

            // Y first for bottom-up building
            for (y in c.corner1.y()..c.corner2.y()) {
                for (x in c.corner1.x()..c.corner2.x()) {
                    for (z in c.corner1.z()..c.corner2.z()) {
                        prepared.add(BlockVector3.at(x, y, z))
                    }
                }
            }
            return prepared
        }
    }

    private fun placeNextBlocks(count: Int): Boolean {
        var placed = 0

        while (placed < count) {
            val index = pointer.incrementAndGet()
            if (index >= blocks.size) return true

            val vec = blocks[index]
            val location = site.worldLocation(vec)

            val blockData = BukkitAdapter.adapt(site.building.getBlock(vec, site.fullRotation)).also {
                rotateBlockData(it, site.fullRotation)
            }

            val currentBlock = site.world.getBlockAt(location)

            // Skip conditions
            when {
                blockData.material == Material.AIR && currentBlock.type == Material.AIR -> continue
                blockData == currentBlock.blockData -> continue
                currentBlock.type in BuildConfig.skipMaterials -> continue
                HookRegistry.sfHook?.isSlimefunBlock(currentBlock) == true -> continue
            }

            // Give drops for replaced blocks
            if (currentBlock.type != blockData.material && currentBlock.type !in BuildConfig.nonDropMaterials) {
                giveDrops(currentBlock)
            }

            // Place block
            ConstructionBlockPlacement.apply(currentBlock, blockData)

            // Effects on first block of tick (only if player is in same world and nearby)
            if (placed == 0 && isPlayerNearby()) {
                playEffects(location, blockData)
            }

            placed++
        }

        return false
    }

    private fun giveDrops(block: org.bukkit.block.Block) {
        for (drop in block.drops) {
            val leftover = site.player.inventory.addItem(drop)
            for (item in leftover.values) {
                site.player.world.dropItem(site.player.location, item)
            }
        }
    }

    private fun playEffects(location: Location, blockData: org.bukkit.block.data.BlockData) {
        // Sound
        if (BuildConfig.playSounds) {
            location.world?.playSound(location, blockData.soundGroup.placeSound, 1f, 1f)
        }

        // Particles
        if (BuildConfig.showParticles) {
            ParticleManager.queue(
                ParticleBuilder(BuildConfig.placeParticle)
                    .count(BuildConfig.particleCount)
                    .location(location)
                    .receivers(listOf(site.player))
                    .offset(BuildConfig.particleOffset, BuildConfig.particleOffset, BuildConfig.particleOffset)
                    .extra(0.05)
            )
        }

        // NPC animations
        if (npcId != -1) {
            HookRegistry.citizensHook?.let { hook ->
                if (ThreadLocalRandom.current().nextDouble() > 0.8) {
                    hook.faceNpc(npcId, location)
                }
                if (blockData.material.isItem) {
                    hook.setMainHand(npcId, ItemStack(blockData.material))
                }
                hook.animateNpc(npcId, CitizensHook.Animation.ARM_SWING)
            }
        }
    }

    // ==================== Utilities ====================

    private fun isPlayerNearby(): Boolean {
        val playerLoc = site.player.location
        return playerLoc.world == site.world &&
            playerLoc.distanceSquared(site.centerBlock) < 2500 // 50^2
    }

    // ==================== Lifecycle ====================

    fun cancel(destroyNpcDelaySeconds: Int) {
        buildTasks.close()
        removeNpcTask?.takeIf { !it.isCancelled }?.cancel()
        removeNpcTask =
            if (destroyNpcDelaySeconds <= 0) {
                destroyNpc()
                null
            } else {
                delayed((destroyNpcDelaySeconds * 20L).ticks) { destroyNpc() }
            }
    }

    fun finishInstantly() {
        cancel(0)
        // Ensure blocks are prepared before placing
        if (blocks.isEmpty()) {
            blocks = prepareBlocks(site)
        }
        placeNextBlocks(1_000_000)
    }
}
